cmd/src/debug_comp.go

```diff
@@ -0,0 +1,255 @@
+package main
+
+import (
+	"archive/zip"
+	"context"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+
+	"golang.org/x/sync/semaphore"
+
+	"github.com/sourcegraph/src-cli/internal/cmderrors"
+)
+
+func init() {
+	usage := `
+'src debug comp' mocks docker cli commands to gather information about a docker-compose Sourcegraph instance.
+
+Usage:
+
+    src debug comp [command options]
+
+Flags:
+
+	-o			Specify the name of the output zip archive.
+	-cfg		Include Sourcegraph configuration json. Defaults to true.
+
+Examples:
+
+    $ src debug comp -o debug.zip
+
+	$ src -v debug comp -cfg=false -o foo.zip
+
+`
+
+	flagSet := flag.NewFlagSet("comp", flag.ExitOnError)
+	var base string
+	var configs bool
+	flagSet.BoolVar(&configs, "cfg", true, "If true include Sourcegraph configuration files. Default value true.")
+	flagSet.StringVar(&base, "o", "debug.zip", "The name of the output zip archive")
+
+	handler := func(args []string) error {
+		if err := flagSet.Parse(args); err != nil {
+			return err
+		}
+
+		//validate out flag
+		if base == "" {
+			return fmt.Errorf("empty -o flag")
+		}
+		// declare basedir for archive file structure
+		var baseDir string
+		if !strings.HasSuffix(base, ".zip") {
+			baseDir = base
+			base = base + ".zip"
+		} else {
+			baseDir = strings.TrimSuffix(base, ".zip")
+		}
+
+		// init context
+		ctx := context.Background()
+		// open pipe to output file
+		out, err := os.OpenFile(base, os.O_CREATE|os.O_RDWR|os.O_EXCL, 0666)
+		if err != nil {
+			fmt.Errorf("failed to open file: %w", err)
+		}
+		defer out.Close()
+		// init zip writer
+		zw := zip.NewWriter(out)
+		defer zw.Close()
+
+		//Gather data for safety check
+		containers, err := getContainers(ctx)
+		if err != nil {
+			fmt.Errorf("failed to get containers for subcommand with err: %v", err)
+		}
+		// Safety check user knows what they are targeting with this debug command
+		log.Printf("This command will archive docker-cli data for %d containers\n SRC_ENDPOINT: %v\n Output filename: %v", len(containers), cfg.Endpoint, base)
+		if verified, _ := verify("Do you want to start writing to an archive?"); !verified {
+			return nil
+		}
+
+		err = archiveCompose(ctx, zw, *verbose, configs, baseDir)
+		if err != nil {
+			return cmderrors.ExitCode(1, nil)
+		}
+		return nil
+	}
+
+	debugCommands = append(debugCommands, &command{
+		flagSet: flagSet,
+		handler: handler,
+		usageFunc: func() {
+			fmt.Println(usage)
+		},
+	})
+}
+
+/*
+Docker functions
+*/
+
+func archiveCompose(ctx context.Context, zw *zip.Writer, verbose, configs bool, baseDir string) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	containers, err := getContainers(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get docker containers: %w", err)
+	}
```

Comment:
You should pass the containers down here since you've already calculated it for the verify check.

Reply:
Do you mean for an informational message? Like archiving container ...

I had experimented with doing an `ls` like output displaying containers that will be archived but decided it was too noisy